### DISCUSSION TITLE
Support for symbology schemas (SLD and similar)

### DISCUSSION THREAD

**Discussion by oyvindi** [openlayers/openlayers#14603] [Ideas]
I find the Openlayers feature styling logic flexible, but also somewhat limited. Some use cases are somewhat impractical by relaying on stylefunction. Some ideas:

- Automatically create styles schemas from an existing SLD (or MapBox style).
- Automatically generate a legend from the schema mentioned above, and perhaps a low level function that generates the legend item descriptors for customization
- Been experimenting with vector tiles recently. We have one use case where we put the map into "night mode", where I do some color manipulation voodoo. The only way to achieve this (as far as I know) is to intercept the stylefunction, and create new instances of style items with altered colors. This implies a lot of instantiations for each pan/zoom. By having a readily generated schema, I could make a clone and mess around with it.

I don't expect a full OGC implementation, but at least the core principles would be nice.

**Comment by ahocevar** [comment]
SLD is probably not going to happen, but Mapbox Style is almost fully supported through the [ol-mapbox-style](https://github.com/openlayers/ol-mapbox-style) library. Are you aware of that? Using that, if you want a night mode, you just apply a different style to the data you have already.

**Reply by oyvindi** [reply]
Yes, I am aware of that library, and that's what being used inside the Openlayers VectorTile implementation. What it does, as far as I can see, is to create a black box stylesheet accessed through a stylefunction. It gives me the style from whatever feature I currently have.

In my case, I'm not interested in creating an entirely new style sheet, I want to clone and modify an existing one. I could of course take the raw MapBox definition, parse it and replace the coloration, but that's a lot of work. The original style sheet is typically provided by external suppliers.

In addition, as mentioned in my post, I can't see a way to generate a legend from that ?


EDIT: ArgGIS Javascript API has the equivalent of what I'm proposing, e.g `UniqueValueRenderer` and `ClassBreaksRenderer`.

**Comment by ahocevar** [comment]
In this case, you might want to take a look at [GeoStyler](https://github.com/geostyler/geostyler). However, ol-mapbox-style also provides utilities for changing styles from the underlying Mapbox Style document. Creating legends is also possible, using [`ol/render/toContext`](https://openlayers.org/en/latest/apidoc/module-ol_render.html) and `layer.getStyleFunction()(legendFeature)`. You can even create a dynamic legend for what's currently in the view, by enabling `recordStyleLayer(true)` in `ol-mapbox-style` and using `layer.getFeaturesInExtent()`.
